src/main/java/core/AbstractPlayer.java

```diff
@@ -52,9 +58,20 @@ public void setForwardModel(AbstractForwardModel model) {
 
     /* Methods that should be implemented in subclass */
 
+    public final AbstractAction getActionWithExploration(AbstractGameState gameState, List<AbstractAction> possibleActions) {
```

Comment:
Small comment: this should be placed above the comment that says 'Methods that should be implemented in subclass' since it's final.

So this is changed so that this method is called by default from Game right? Can we call this instead 'getAction' to avoid confusion ('why are we getting actions with exploration all the time?'), and make the other one private '_getAction' if it's not accessed directly from outside anymore?

This function is also missing documentation, which would be helpful here for the future :)

Reply:
Will add documentation!

I dithered about the naming convention - an went with getActionWithExploration() for two reasons:
1) To avoid having to change any existing agents, which all have to implement getAction().
2) And because this is always called from within the final method, so the name would be descriptive, but invisible to the casual implementor.

But 1) is resolvable with some adroit use of Intellij refactoring - so consider it done.